pkg/controller/direct/cloudidentity/group_controller.go

```diff
@@ -0,0 +1,274 @@
+}
+
+func NewGroupModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &modelGroup{config: *config}, nil
+}
+
+var _ directbase.Model = &modelGroup{}
+
+type modelGroup struct {
+	config config.ControllerConfig
+}
+
+func (m *modelGroup) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.CloudIdentityGroup{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Handle TF default values
+	if obj.Spec.InitialGroupConfig == nil {
+		obj.Spec.InitialGroupConfig = direct.LazyPtr("EMPTY")
+	}
+
+	id, err := krm.NewGroupIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get cloudidentitygroup GCP client
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	gcpClient, err := gcp.NewService(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return &GroupAdapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func (m *modelGroup) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type GroupAdapter struct {
+	id        *krm.GroupIdentity
+	gcpClient *gcp.Service
+	desired   *krm.CloudIdentityGroup
+	actual    *cloudidentitygrouppb.Group
+}
+
+var _ directbase.Adapter = &GroupAdapter{}
+
+// Find retrieves the GCP resource.
+// Return true means the object is found. This triggers Adapter `Update` call.
+// Return false means the object is not found. This triggers Adapter `Create` call.
+// Return a non-nil error requeues the requests.
+func (a *GroupAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting Group", "name", a.id)
+
+	// Check whether Config Connector knows the resource identity.
+	// If not, Config Connector saves one GCP GET call, and starts the CREATE call directly.
+	// This is mostly for GCP services that do not allow user to specify ID, but assign an ID when creating the object.
+	if a.id.ID() == "" {
+		return false, nil
+	}
+
+	resource, err := a.gcpClient.Groups.Get(a.id.String()).Context(ctx).Do()
+	if err != nil {
+		return false, fmt.Errorf("getting Group %q: %w", a.id, err)
+	}
+
+	a.actual = convertAPIToProto(resource)
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *GroupAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating Group", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := CloudIdentityGroupSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := convertProtoToAPI(resource)
```

Comment:
Extra layer of conversion seems unnecessary, is it possible that we manually fix the existing mapper functions to map directly between KRM and API?